### PR TITLE
Explicitly creating method to copy http headers to internal metadata

### LINKED ISSUE #1730: Explicitly creating method to copy http headers to internal metadata
No bugs but for better maintainability/readability - explicitly creating method to copy http headers to internal metadata.

> /area runtime

## What version of Dapr?

0.8.0

## Expected Behavior

Explicitly creating method to copy http headers to internal metadata.

## Actual Behavior

Currently GrpcMetadataToInternalMetadata is being used for http headers.

### PR DESCRIPTION
**Pull request by shalabhms**
# Description

Explicitly creating method to copy http headers to internal metadata

## Issue reference

Closes #1730 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/messaging/v1/invoke_method_request.go [hunk 1]
```diff
@@ -66,7 +67,17 @@ func (imr *InvokeMethodRequest) WithActor(actorType, actorID string) *InvokeMeth
 
 // WithMetadata sets metadata
 func (imr *InvokeMethodRequest) WithMetadata(md map[string][]string) *InvokeMethodRequest {
-	imr.r.Metadata = GrpcMetadataToInternalMetadata(md)
+	imr.r.Metadata = MetadataToInternalMetadata(md)
+	return imr
+}
+
+// WithFastHTTPHeaders sets fasthttp request headers
+func (imr *InvokeMethodRequest) WithFastHTTPHeaders(header *fasthttp.RequestHeader) *InvokeMethodRequest {
```
Comment (yaron2) [inline pkg/messaging/v1/invoke_method_request.go:75]: Please make this a non-pointer.

**Comment by shalabhms** [comment]
@youngbupark , I have exposed WithFastHttpHeaders with MetadataToInternalMetadata as common one. Let me know  your opinion , we can continue to have GrpcMetadataToInternalMetadata and  add "MetadataToInternalMetadata" logic for http headers directly in WithFastHttpHeaders method. Currently only caveat is that its checking for -bin headers which it is already doing though..

**Comment by youngbupark** [comment]
> @youngbupark , I have exposed WithFastHttpHeaders with MetadataToInternalMetadata as common one. Let me know your opinion , we can continue to have GrpcMetadataToInternalMetadata and add "MetadataToInternalMetadata" logic for http headers directly in WithFastHttpHeaders method. Currently only caveat is that its checking for -bin headers which it is already doing though..

no objection :) we already discussed it.

**Comment by shalabhms** [comment]
> @youngbupark , I have exposed WithFastHttpHeaders with MetadataToInternalMetadata as common one. Let me know your opinion , we can continue to have GrpcMetadataToInternalMetadata and add "MetadataToInternalMetadata" logic for http headers directly in WithFastHttpHeaders method. Currently only caveat is that its checking for -bin headers which it is already doing though..



> > @youngbupark , I have exposed WithFastHttpHeaders with MetadataToInternalMetadata as common one. Let me know your opinion , we can continue to have GrpcMetadataToInternalMetadata and add "MetadataToInternalMetadata" logic for http headers directly in WithFastHttpHeaders method. Currently only caveat is that its checking for -bin headers which it is already doing though..
> 
> no objection :) we already discussed it.

ok..I pushed though another commit..for completely separating out. I will revert that commit then..

File: pkg/messaging/v1/invoke_method_request.go (hunk 1 above)
Comment (shalabhms) [reply pkg/messaging/v1/invoke_method_request.go:75]: Copying headers in fasthttp is not allowed. 
// Copying Header by value is forbidden. Use pointer to Header instead.

Also I noticed that we are doing through pointer in response code path : https://github.com/dapr/dapr/blob/a63ada63a11c86c51f4aa36c4b081add677af544/pkg/messaging/v1/invoke_method_response.go#L69

Comment (shalabhms) [reply pkg/messaging/v1/invoke_method_request.go:75]: @yaron2 , @youngbupark , we can not copy headers so need to use pointer way.

Comment (yaron2) [reply pkg/messaging/v1/invoke_method_request.go:75]: I find it odd that the `Request` struct doesn't hold the `RequestHeader` as a pointer as it does with many other types.

Comment (youngbupark) [reply pkg/messaging/v1/invoke_method_request.go:75]: @shalabhms Can we pass requestCtx pointer instead of fasthttp.ResponseHeader?

Comment (shalabhms) [reply pkg/messaging/v1/invoke_method_request.go:75]: @youngbupark , as you approved now, I am resolving this conversation.  
@yaron2 , can you confirm also ?

**Pull request merged**